### PR TITLE
Create new test file for testing resize OSD

### PR DESCRIPTION
**Pull request by yitzhak12**
The PR implements the basic test scenario of the resize osd feature. For more details about the feature, see https://issues.redhat.com/browse/RHSTOR-5251, and for the test details, see the test plan spreadsheet: https://docs.google.com/spreadsheets/d/1wCSwMk_L4SikzZ2UarUH1-VCXyGkHs-ZNdDEtWPG7f4/edit#gid=105186681&range=B8. 
In the test, we will perform the following:

- Prepare all the relevant data before the resizing(Save the old pods, PVCs, PVs, create pods and PVCs and run IO, save the file md5sum). 
- Resize the osd by multiplying 2(This will be done by patching the storagecluster storage param).
- Check that the resources are in the expected state and that the new size is reflected in the resources and Ceph. Also, check that the md5sum of the files didn't change after the resize osd. 
- Check the cluster and Ceph health

### CODE REVIEW

**Review by fbalak — COMMENTED** [review]
The functionality looks good but I have 2 suggestions:

 - New `*post_resize_osd ` functions shouldn't be in cluster.py module in my opinion. It would be better to have a new module (e.g. `osd_resize.py`) that would contain those checks and resize functionality.  Some of those functions use storagecluster cr but their functionality is focused on osds and resizing.

 - It is better to raise an `Exception` in module function instead of `assert`. `Asserts` should be used in tests but functions that could be reused on multiple places should raise an `Exception` that can be handled (e.g. negative scenario where you want it to fail).

**Comment by yitzhak12** [comment]
Thanks @fbalak. I will modify the code accordingly.

File: ocs_ci/ocs/resources/osd_resize.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        condition=constants.STATUS_RUNNING,
+        selector=constants.OSD_APP_LABEL,
+        resource_count=old_osd_pods_count,
+        timeout=300,
+        sleep=20,
+    )
+
+    logger.info(
+        f"Check that the number of the new OSD pods are exactly {old_osd_pods_count}"
+    )
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/osd_resize.py:78]: can we also add a validation that the Running OSD pod names are different than old_osd_pods names?

File: ocs_ci/ocs/resources/osd_resize.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        "Check that the PVC and PV sizes are equal to the expected storage size"
+    )
+    current_osd_pvcs = get_deviceset_pvcs()
+    expected_storage_size_in_gb = convert_device_size(expected_storage_size, "GB")
+    pvc_sizes = [pvc.size for pvc in current_osd_pvcs]
+    logger.info(f"PVC sizes = {pvc_sizes}")
+    if not all([p_size == expected_storage_size_in_gb for p_size in pvc_sizes]):
+        raise StorageSizeNotReflectedException(
+            f"The PVC sizes are not equal to the expected storage size {expected_storage_size_in_gb}"
+        )
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/osd_resize.py:158]:
```suggestion
        raise StorageSizeNotReflectedException(
            f"The PVC sizes are not equal to the expected storage size {expected_storage_size_in_gb}. "
            f"PVC sizes: {[p_size for p_size in pvc_sizes]}
        )
```

File: ocs_ci/ocs/resources/osd_resize.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    current_osd_pvs = get_deviceset_pvs()
+    pv_sizes = [get_pv_size(pv.get()) for pv in current_osd_pvs]
+    logger.info(f"PV sizes {pv_sizes}")
+    if not all([p_size == expected_storage_size_in_gb for p_size in pv_sizes]):
+        raise StorageSizeNotReflectedException(
+            f"The PV sizes are not equal to the expected storage size {expected_storage_size_in_gb}"
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/osd_resize.py:165]:
```suggestion
            f"The PV sizes are not equal to the expected storage size {expected_storage_size_in_gb}. "
            f"PV sizes: {[p_size for p_size in pv_sizes]}"
```

File: ocs_ci/ocs/resources/osd_resize.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    logger.info("Check the Ceph state post resize OSD")
+    check_ceph_state_post_resize_osd()
+    logger.info("All the Ceph verification steps post resize osd finished successfully")
+
+
+def ceph_verification_steps_post_resize_osd(
+    old_osd_pods, old_osd_pvcs, old_osd_pvs, expected_storage_size, num_of_tries=6
+):
+    """
+    Try to execute the function 'base_ceph_verification_steps_post_resize_osd' a number of tries
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/osd_resize.py:238]: why are we doing that?

File: ocs_ci/ocs/resources/storage_cluster.py [hunk 5]
```diff
@@ -2669,3 +2650,56 @@ def validate_serviceexport():
...
+    Resize the OSD(e.g., from 512 to 1024, 1024 to 2048, etc.)
+
+    Args:
+        new_osd_size (str): The new osd size(e.g, 512Gi, 1024Gi, 1Ti, 2Ti, etc.)
+
+    Returns:
+        bool: True in case if changes are applied. False otherwise
+
+    """
+    sc = get_storage_cluster()
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/storage_cluster.py:2696]: please add a check that the new requested size is greater than the existing OSD size.
And should we add a test case of tyring to resize to a smaller size?

File: ocs_ci/ocs/resources/osd_resize.py (hunk 1 above)
Comment (yitzhak12) [reply ocs_ci/ocs/resources/osd_resize.py:78]: Yes, I can add it.

File: ocs_ci/ocs/resources/osd_resize.py (hunk 2 above)
Comment (yitzhak12) [reply ocs_ci/ocs/resources/osd_resize.py:158]: Here, it logs the PVC sizes anyway a few lines back. Should we still add this log message?

File: ocs_ci/ocs/resources/osd_resize.py (hunk 3 above)
Comment (yitzhak12) [reply ocs_ci/ocs/resources/osd_resize.py:165]: Here, it logs the PV sizes anyway a few lines back. Should we still add this log message?

File: ocs_ci/ocs/resources/osd_resize.py (hunk 4 above)
Comment (yitzhak12) [reply ocs_ci/ocs/resources/osd_resize.py:238]: Yes, right. This is regarding the comment here: https://bugzilla.redhat.com/show_bug.cgi?id=2257497#c14.

> 4. If the resize did not take effect yet, restart the OSD pods again, possibly even multiple times to get the resize to take effect. (But if it's AWS and the OSDs have already been resized previously, the resize may not be possible according to initial testing.)

File: ocs_ci/ocs/resources/storage_cluster.py (hunk 5 above)
Comment (yitzhak12) [reply ocs_ci/ocs/resources/storage_cluster.py:2696]: Okay, np. 
We added a negative test in the test plan: https://docs.google.com/spreadsheets/d/1wCSwMk_L4SikzZ2UarUH1-VCXyGkHs-ZNdDEtWPG7f4/edit#gid=105186681&range=B24

File: ocs_ci/ocs/resources/osd_resize.py (hunk 2 above)
Comment (ebenahar) [reply ocs_ci/ocs/resources/osd_resize.py:158]: ack, then no need

File: ocs_ci/ocs/resources/osd_resize.py (hunk 4 above)
Comment (ebenahar) [reply ocs_ci/ocs/resources/osd_resize.py:238]: ack, thanks for the reference

File: ocs_ci/ocs/resources/storage_cluster.py (hunk 5 above)
Comment (ebenahar) [reply ocs_ci/ocs/resources/storage_cluster.py:2696]: do you plan to add it in a follow up PR?

Comment (yitzhak12) [reply ocs_ci/ocs/resources/storage_cluster.py:2714]: I have implemented the osd size check. 
I will add the negative test in another PR. I need to check first manually what is the behavior

File: ocs_ci/ocs/resources/osd_resize.py (hunk 1 above)
Comment (yitzhak12) [reply ocs_ci/ocs/resources/osd_resize.py:78]: Done

**Review by ebenahar — APPROVED** [review]

**Review by ebenahar — APPROVED** [review]

**Review by fbalak — APPROVED** [review]

**Comment by petr-balogh** [comment]
If this PR is going to be cherry-picked we need to include fix from Vijay:
https://github.com/red-hat-storage/ocs-ci/pull/9678

As this code braking the add capacity test:
```
>       device_sets_required = int(osd_size_capacity_requested / osd_size_existing)
E       TypeError: unsupported operand type(s) for /: 'str' and 'str'

ocs_ci/ocs/resources/storage_cluster.py:1470: TypeError
```
@yitzhak12  FYI

**Pull request merged**